library/src/main/java/com/datatorrent/lib/join/POJOInnerJoinOperator.java

```diff
@@ -0,0 +1,247 @@
+{
+  private static final transient Logger LOG = LoggerFactory.getLogger(POJOInnerJoinOperator.class);
+  private transient long timeIncrement;
+  private transient FieldObjectMap[] inputFieldObjects = (FieldObjectMap[])Array.newInstance(FieldObjectMap.class, 2);
+  protected Class<?> outputClass;
+  private long time = System.currentTimeMillis();
+
+  @OutputPortFieldAnnotation(schemaRequired = true)
+  public final transient DefaultOutputPort<Object> outputPort = new DefaultOutputPort<Object>()
+  {
+    @Override
+    public void setup(Context.PortContext context)
+    {
+      outputClass = context.getValue(Context.PortContext.TUPLE_CLASS);
+    }
+  };
+
+  @InputPortFieldAnnotation(schemaRequired = true)
+  public transient DefaultInputPort<Object> input1 = new DefaultInputPort<Object>()
+  {
+    @Override
+    public void setup(Context.PortContext context)
+    {
+      inputFieldObjects[0].inputClass = context.getValue(Context.PortContext.TUPLE_CLASS);
+    }
+
+    @Override
+    public void process(Object tuple)
+    {
+      processTuple(tuple,true);
+    }
+
+    @Override
+    public StreamCodec<Object> getStreamCodec()
+    {
+      return getInnerJoinStreamCodec(true);
+    }
+  };
+
+  @InputPortFieldAnnotation(schemaRequired = true)
+  public transient DefaultInputPort<Object> input2 = new DefaultInputPort<Object>()
+  {
+    @Override
+    public void setup(Context.PortContext context)
+    {
+      inputFieldObjects[1].inputClass = context.getValue(Context.PortContext.TUPLE_CLASS);
+    }
+
+    @Override
+    public void process(Object tuple)
+    {
+      processTuple(tuple,false);
+    }
+
+    @Override
+    public StreamCodec<Object> getStreamCodec()
+    {
+      return getInnerJoinStreamCodec(false);
+    }
+  };
+
+  @Override
+  public void setup(Context.OperatorContext context)
+  {
+    timeIncrement = context.getValue(Context.OperatorContext.APPLICATION_WINDOW_COUNT) *
+      context.getValue(Context.DAGContext.STREAMING_WINDOW_SIZE_MILLIS);
+    super.setup(context);
+    for (int i = 0; i < 2; i++) {
+      inputFieldObjects[i] = new FieldObjectMap();
+    }
+  }
+
+  /**
+   * Extract the time value from the given tuple
+   * @param tuple given tuple
+   * @param isStream1Data Specifies whether the given tuple belongs to stream1 or not.
+   * @return
+   */
+  @Override
+  public long extractTime(Object tuple, boolean isStream1Data)
+  {
+    return timeFields == null ? time : (long)(isStream1Data ? inputFieldObjects[0].timeFieldGet.get(tuple) :
+          inputFieldObjects[1].timeFieldGet.get(tuple));
+  }
+
+  /**
+   * Create getters for the key and time fields and setters for the include fields.
+   */
+  protected void generateSettersAndGetters()
+  {
+    for (int i = 0; i < 2; i++) {
+      Class inputClass = inputFieldObjects[i].inputClass;
+      try {
+        Class c = ClassUtils.primitiveToWrapper(inputClass.getField(keyFields.get(i)).getType());
```

Comment:
keyFields is populated in setup method of the operator but this method is called from activate? activate of operator is called before setup.
Can you please check if this is giving desired result? Am I missing something?